### PR TITLE
Some fixes to the linear algebra functions

### PR DESCRIPTION
**Pull request by asmeurer**

### CODE REVIEW

**Comment by asmeurer** [comment]
As I noted on Slack, the broadcasting rule for matmul is wrong too (the last bullet point). It should not broadcast the "matrix dimensions". But I'm having a hard time figuring out a way to rewrite it that isn't too verbose. Any suggestions are welcome.

File: spec/API_specification/array_object.md [hunk 1]
```diff
@@ -837,7 +837,9 @@ The `matmul` function must implement the same semantics as the built-in `@` oper
 #### Raises
 
 -   if either `self` or `other` is a zero-dimensional array.
--   if `self` is a one-dimensional array having shape `(N)`, `other` is a one-dimensional array having shape `(M)`, and `N != M`.
+-   if `self` is a one-dimensional array having shape `(K)`, `other` is a one-dimensional array having shape `(L)`, and `K != L`.
+-   if `self` is a one-dimensional array having shape `(K)`, `other` is an array having shape `(..., L, M)`, and `K != L`.
+-   if `self` is an array having shape `(..., N, K)`, `other` is a one-dimensional array having shape `(L)`, and `K != L`.
```
Comment (kgryte) [inline spec/API_specification/array_object.md:842]:
```suggestion
-   if `self` is an array having shape `(..., M, K)`, `other` is a one-dimensional array having shape `(L)`, and `K != L`.
```

File: spec/API_specification/array_object.md [hunk 2]
```diff
@@ -837,7 +837,9 @@ The `matmul` function must implement the same semantics as the built-in `@` oper
 #### Raises
 
 -   if either `self` or `other` is a zero-dimensional array.
--   if `self` is a one-dimensional array having shape `(N)`, `other` is a one-dimensional array having shape `(M)`, and `N != M`.
+-   if `self` is a one-dimensional array having shape `(K)`, `other` is a one-dimensional array having shape `(L)`, and `K != L`.
+-   if `self` is a one-dimensional array having shape `(K)`, `other` is an array having shape `(..., L, M)`, and `K != L`.
```
Comment (kgryte) [inline spec/API_specification/array_object.md:841]:
```suggestion
-   if `self` is a one-dimensional array having shape `(K)`, `other` is an array having shape `(..., L, N)`, and `K != L`.
```

File: spec/API_specification/linear_algebra_functions.md [hunk 3]
```diff
@@ -52,7 +52,9 @@ The `matmul` function must implement the same semantics as the built-in `@` oper
 #### Raises
 
 -   if either `x1` or `x2` is a zero-dimensional array.
--   if `x1` is a one-dimensional array having shape `(N)`, `x2` is a one-dimensional array having shape `(M)`, and `N != M`.
+-   if `x1` is a one-dimensional array having shape `(K)`, `x2` is a one-dimensional array having shape `(L)`, and `K != L`.
+-   if `x1` is a one-dimensional array having shape `(K)`, `x2` is an array having shape `(..., L, M)`, and `K != L`.
+-   if `x1` is an array having shape `(..., N, K)`, `x2` is a one-dimensional array having shape `(L)`, and `K != L`.
```
Comment (kgryte) [inline spec/API_specification/linear_algebra_functions.md:57]:
```suggestion
-   if `x1` is a one-dimensional array having shape `(K)`, `x2` is an array having shape `(..., L, N)`, and `K != L`.
-   if `x1` is an array having shape `(..., M, K)`, `x2` is a one-dimensional array having shape `(L)`, and `K != L`.
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @asmeurer!

**Comment by kgryte** [comment]
Re: `matmul`. Will submit a follow-up PR to address broadcasting.

**Pull request merged**